File: memento/src/test/java/com/alexstyl/specialdates/events/peopleevents/PeopleDynamicNamedaysProviderTest.java
package com.alexstyl.specialdates.events.peopleevents;

import com.alexstyl.specialdates.contact.Contact;
import com.alexstyl.specialdates.contact.ContactFixture;
import com.alexstyl.specialdates.contact.ContactsProvider;
import com.alexstyl.specialdates.date.ContactEvent;
import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.date.TimePeriod;
import com.alexstyl.specialdates.events.namedays.NamedayLocale;
import com.alexstyl.specialdates.events.namedays.NamedayUserSettings;
import com.alexstyl.specialdates.events.namedays.calendar.NamedayCalendar;
import com.alexstyl.specialdates.events.namedays.calendar.OrthodoxEasterCalculator;
import com.alexstyl.specialdates.events.namedays.calendar.resource.NamedayCalendarProvider;
import com.alexstyl.specialdates.events.namedays.calendar.resource.TestNamedayCalendarBuilder;

import java.util.ArrayList;
import java.util.List;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static org.fest.assertions.api.Assertions.assertThat;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class PeopleDynamicNamedaysProviderTest {

    private PeopleDynamicNamedaysProvider calculator;

    private static final NamedayLocale LOCALE = NamedayLocale.GREEK;
    private static final OrthodoxEasterCalculator calculator1 = new OrthodoxEasterCalculator();
    private final Contact EASTER_CELEBRATING_CONTACT = ContactFixture.aContactCalled("Λάμπρος");
    private static final int YEAR = 2016;
    @Mock
    private NamedayCalendarProvider namedayCalendarProvider;
    @Mock
    private NamedayUserSettings mockSettings;
    @Mock
    private ContactsProvider mockContactsProvider;

    @Before
    public void setUp() {
        NamedayCalendar namedayCalendar = new TestNamedayCalendarBuilder()
                .forLocale(LOCALE)
                .forYear(YEAR)
                .build();

        when(namedayCalendarProvider.loadNamedayCalendarForLocale(any(NamedayLocale.class), any(Integer.class))).thenReturn(namedayCalendar);
        when(mockSettings.getSelectedLanguage()).thenReturn(LOCALE);
        when(mockSettings.isEnabled()).thenReturn(true);
        calculator = new PeopleDynamicNamedaysProvider(mockSettings, namedayCalendarProvider, mockContactsProvider);

    }

    @Test
    public void gettingSpecialNamedaysOnSpecificDateOnlyReturnsTheEventsForThatDate() {
        List<Contact> testContacts = createSomeContacts();
        testContacts.add(EASTER_CELEBRATING_CONTACT);
        when(mockContactsProvider.getAllContacts()).thenReturn(testContacts);

        Date easterDate = calculator1.calculateEasterForYear(YEAR);
        List<ContactEvent> contactEvents = calculator.fetchEventsBetween(TimePeriod.Companion.between(easterDate, easterDate));
        assertThat(contactEvents).hasSize(1);
        assertThat(contactEvents.get(0).getContact()).isEqualTo(EASTER_CELEBRATING_CONTACT);
    }

    private List<Contact> createSomeContacts() {
        ArrayList<Contact> contacts = new ArrayList<>();
        contacts.add(ContactFixture.aContactCalled("Αβδηρος"));
        contacts.add(ContactFixture.aContactCalled(("Αγις")));
        return contacts;
    }

}

File: memento/src/test/java/com/alexstyl/specialdates/events/peopleevents/PeopleDynamicNamedaysProviderTest.kt
package com.alexstyl.specialdates.events.peopleevents

import com.alexstyl.specialdates.contact.Contact
import com.alexstyl.specialdates.contact.ContactFixture
import com.alexstyl.specialdates.contact.Contacts
import com.alexstyl.specialdates.contact.ContactsProvider
import com.alexstyl.specialdates.contact.ContactsProviderSource
import com.alexstyl.specialdates.date.TimePeriod
import com.alexstyl.specialdates.events.namedays.NamedayLocale
import com.alexstyl.specialdates.events.namedays.NamedayUserSettings
import com.alexstyl.specialdates.events.namedays.calendar.OrthodoxEasterCalculator
import com.alexstyl.specialdates.events.namedays.calendar.resource.NamedayCalendarProvider
import com.alexstyl.specialdates.events.namedays.calendar.resource.TestNamedayCalendarBuilder
import org.fest.assertions.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.BDDMockito.given
import org.mockito.Matchers.any
import org.mockito.Mock
import org.mockito.runners.MockitoJUnitRunner

@RunWith(MockitoJUnitRunner::class)
class PeopleDynamicNamedaysProviderTest {

    private lateinit var calculator: PeopleDynamicNamedaysProvider
    @Mock
    private lateinit var namedayCalendarProvider: NamedayCalendarProvider
    @Mock
    private lateinit var mockSettings: NamedayUserSettings
    @Mock
    private lateinit var mockSource: ContactsProviderSource
    private val EASTER_CELEBRATING_CONTACT = ContactFixture.aContactCalled("Λάμπρος")

    @Before
    fun setUp() {
        val namedayCalendar = TestNamedayCalendarBuilder()
                .forLocale(LOCALE)
                .forYear(YEAR)
                .build()

        given(namedayCalendarProvider.loadNamedayCalendarForLocale(any(NamedayLocale::class.java), any(Int::class.java))).willReturn(namedayCalendar)
        given(mockSettings.selectedLanguage).willReturn(LOCALE)
        given(mockSettings.isEnabled).willReturn(true)
        calculator = PeopleDynamicNamedaysProvider(mockSettings, namedayCalendarProvider,
                ContactsProvider(mapOf(Pair(1, mockSource)))
        )

    }

    @Test
    fun gettingSpecialNamedaysOnSpecificDateOnlyReturnsTheEventsForThatDate() {
        val testContacts = createSomeContacts()
        testContacts.add(EASTER_CELEBRATING_CONTACT)
        given(mockSource.allContacts).willReturn(Contacts(1, testContacts))

        val orthodoxEasterCalculator = OrthodoxEasterCalculator()
        val easterDate = orthodoxEasterCalculator.calculateEasterForYear(YEAR)
        val contactEvents = calculator.fetchEventsBetween(TimePeriod.between(easterDate, easterDate))
        assertThat(contactEvents).hasSize(1)
        assertThat(contactEvents[0].contact).isEqualTo(EASTER_CELEBRATING_CONTACT)
    }

    private fun createSomeContacts(): MutableList<Contact> = arrayListOf(
            ContactFixture.aContactCalled("Αβδηρος"),
            ContactFixture.aContactCalled("Αγις")
    )

    companion object {

        private val LOCALE = NamedayLocale.GREEK

        private const val YEAR = 2016
    }

}
